### DISCUSSION TITLE
Failure of soup orientation (octahedron)

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#6606] [Q&A] [answered]
Hello,

 I have an octahedron with these vertices (phi is the golden ratio):

```
( 0.0 , 0.0 , phi )
( 0.0 , phi , 0.0 )
( phi , 0.0 , 0.0 )
( 0.0 , 0.0 , -phi )
( 0.0 , -phi , 0.0 )
( -phi , 0.0 , 0.0 )
```

Its faces are ill-oriented. So I orient them with CGAL, and the result is not correct.
However if I take the face indices given by this result and I use these vertices instead:

```
( 0.0 , 0.0 , -phi )
( 0.0 , -phi , 0.0 )
( -phi , 0.0 , 0.0 )
( 0.0 , 0.0 , phi )
( 0.0 , phi , 0.0 )
( phi , 0.0 , 0.0 )
```

then the result is correct.

I will give you the code tomorrow, I'm too tired now. I just posted not to forget.

**Comment by stla** [comment]
[Here is the code](https://gist.github.com/stla/097d4c3b090bf22ac60133cecd990367).

**Comment by sloriot** [comment]
The code works without issue with master, CGAL 5.4 and CGAL 5.3. I got `octahedron.off` written. When `false` is returned it is most of the time due to a non-manifold vertex or a set of edge cutting a non-orientable surface (like mobius strip for example). In this case there is no reason it would be `false`.

**Reply by stla** [reply]
I know the code works without issue. I'm just saying the faces it returns are not correctly oriented. Wait, I will show you some pictures.

**Reply by sloriot** [reply]
Looking at the mesh in the demo they seems fine to me. The mesh is outward oriented, free of self-intersection and closed.

**Reply by stla** [reply]
Look. The green octahedron is the one obtained with CGAL. It is sick ;-)

```r
library(rgl) # for plotting
library(MeshesOperations) # wrapper to CGAL

phi <- (1+sqrt(5))/2
vertices <- rbind(
  c( 0.0 , 0.0 , phi )
  , c( 0.0 , phi , 0.0 )
  , c( phi , 0.0 , 0.0 )
  , c( 0.0 , 0.0 , -phi )
  , c( 0.0 , -phi , 0.0 )
  , c( -phi , 0.0 , 0.0 )
)

faces <- rbind(
  c(0L, 1L, 2L), # ('L' is for integers)
  c(0L, 4L, 5L), 
  c(5L, 0L, 1L),
  c(2L, 0L, 4L),
  c(3L, 2L, 1L),
  c(3L, 5L, 1L),
  c(3L, 2L, 4L),
  c(3L, 5L, 4L)
) + 1L # add one because R indices start at 1

# we have our polygon soup, let's orient it with CGAL now:
mesh <- Mesh(vertices, faces)

# let's plot this mesh now, the front of the faces filled in green, 
# and the back of the faces as wireframed polygons
rglmesh <- tmesh3d(
  vertices = t(mesh$vertices),
  indices = t(mesh$faces),
  homogeneous = FALSE
)
open3d()
view3d(-5, 15, zoom = 0.7)
shade3d(rglmesh, front = "fill", back = "lines", lwd = 4, color = "green")
# we see the lines, that shows the faces are badly oriented
# not convinced ? let's compare
# here is an octahedron whose faces are correctly oriented:
open3d()
view3d(-5, 15, zoom = 0.7)
shade3d(octahedron3d(), front = "fill", back = "lines", lwd = 4, color = "red")
```

![octahedra](https://user-images.githubusercontent.com/4466543/170127174-7e8da1f6-d832-44e4-a870-dfa656dfada6.png)

**Reply by stla** [reply]
And if I negate the vertices (which comes down to changing their order here), this becomes correct:

```r
rglmesh2 <- tmesh3d(
  vertices = -t(mesh$vertices),
  indices = t(mesh$faces),
  homogeneous = FALSE
)
open3d()
view3d(-5, 15, zoom = 0.7)
shade3d(rglmesh2, front = "fill", back = "lines", lwd = 4, color = "cyan")
```

![octahedron](https://user-images.githubusercontent.com/4466543/170128037-3fb3b9ba-ea97-481c-b80a-bf9b7169bd12.png)

**Reply by stla** [reply]
Note that I didn't take the same face indices as in my C++ code. The mesh I get with CGAL is *inward* oriented. So ok, this is consistent (I mean there's no mix of outward and inward), but this is inward, not outward.

**Comment by sloriot** [comment, marked as answer]
`orient_polygon_soup()` does not do anything about the inward / outward orientation. You can call `PMP::is_outward_oriented` and `PMP::reverse_face_orientations()` if needed.
